lib/frameworks/jasmine.py

```diff
@@ -1,11 +1,102 @@
-import json, retrying, urllib
+import json, retrying, urllib, sys
 from selenium.webdriver.support.ui import WebDriverWait
 from itertools import groupby
 from xml.etree import ElementTree
 from lib import log
 
 webDriverWaitTimeout = 300
 
+from concurrent import futures
+
+def runTestsInParallel( drivers, timeout ):
+
+    log.write( "Calculating number of tests ... " )
+    tests = getTests( drivers[0]["driver"], drivers[0]["testsUrl"] )
+    log.writeln( "%d tests found" % len( tests ) )
+
+    test_results = runTestsInPoolDrivers( drivers, tests )
+    testResults = test_results['results']
+    retries_tests = test_results['retries_tests']
+
+    retries = 5
+
+    while retries_tests != [] and retries:
+
+      log.writeln( 'Retries tests: %s' % str(retries_tests) )
+
+      retries -= 1
+
+      retries_test_results = runTestsInPoolDrivers( drivers, retries_tests )
+
+      for retries_test_result in retries_test_results['results']:
+        for testResult in testResults:
+          if urllib.unquote( testResult['test_url'].split("?spec=")[1] ) == urllib.unquote( retries_test_result['test_url'].split("?spec=")[1] ):
+            testResult['json'] = retries_test_result['json']
+            testResult['junit'] = retries_test_result['junit']
+
+      retries_tests = retries_test_results['retries_tests']
+
+    suites = groupTestSuites( map( lambda x: x[ "json" ], testResults ) )
+    xmlSuites = map( lambda x: ElementTree.tostring( x ), groupXmlSuites( map( lambda x: x[ "junit" ], testResults ) ) )
+
+    testResults = {
+
+        "json": {
+
+            "suites": suites,
+            "durationSec": sum( map( lambda x: x[ "durationSec" ], suites ) ),
+            "passed": all( map( lambda x: x[ "passed" ], suites ) )
+        },
+
+        "junit": '<?xml version="1.0" encoding="UTF-8" ?>\n<testsuites>\n%s\n</testsuites>\n' % "\n".join( xmlSuites )
+    }
+
+    return testResults
+
+def runTestsInPoolDrivers( drivers, tests ):
+
+    test_results = {'results':[], 'retries_tests':[]}
+    counter = 0
+
+    with futures.ThreadPoolExecutor( max_workers=len( drivers ) ) as executor:
+      executions = []
+      for test in tests:
+          executions.append( executor.submit( runTestByDriversPool, drivers, test, 60 ) )
+      for execution in executions:
+          exception = execution.exception()
+          counter += 1
+          if exception is not None:
+              log.writeln( "Got exception that broke everything: " + str( exception ) + "\n" )
+          else:
+              testResult = execution.result()
+              test = urllib.unquote(testResult['test_url'].split("?spec=")[1])
+              log.write( "test %d: %s ... " % ( counter, test ) )
+
+              if not isPassed( testResult[ "json" ] ):
+                log.writeln( "failed" )
+                log.writeln( str(testResult) )
+                test_results['retries_tests'].append( test )
```

Comment:
This change overcomplicated retry logic. You do not need to store `retried_tests`. I believe you should use `queue` instead of `list` to keep all executions and while there is something in queue you should dequeue from executions and enqueue failed ones. It is even better to just retry failed tests inside runTestByDriversPool method if possible.